Added CSV download

PrzemyslawSarnacki opened the pull request:
I've added CSV download support for reports. I've tried not to repeat myself as much as it was reasonably possible. Due to use of polish decimal separator (which is comma instead of dot) csv files have semicolon (;) as a delimiter.  I'm looking forward for review of my implementation.

Kwpolska commented on .gitignore:
```diff
@@ -42,5 +42,9 @@ nosetests.xml
 # Node
 node_modules
 
+DjangoExpenses/
```
Those `.gitignore` changes are not relevant to `django-expenses`, please undo them.

Kwpolska commented on expenses/reports.py:
```diff
@@ -149,14 +149,40 @@ def tabulate(self, results: typing.Iterable, engine: Engine) -> SafeString:
+            raise ValueError("Results do not match expected column headers")
+        
+
+        return mark_safe(
```
This function should return a regular string and create a CSV file using the `csv` library, not try produce it by hand with string templating operations. Things get more complex if a `"`, `\`, or newline appears in fields (and it’s possible with free-form inputs, such as the category names.

Also, note that the CSV file produced contains the currency symbol. This isn’t usual in CSV files, not all software will interpret it correctly. Microsoft Excel could in theory interpret it, but it expects a regular space (U+0020) instead of a non-breaking space (U+00A0) that we use. The rendering pipeline assumes HTML a bit too much; this could be fixed by injecting `format_money` etc. as an argument (and pass a different function depending on output format) + an `is-html` argument to avoid creating links in `VendorStats`.

Kwpolska commented on expenses/reports.py:
```diff
@@ -241,6 +267,16 @@ def get_column_headers(self, engine: Engine) -> (typing.List[str], typing.List[s
         elif self.query_type == "month":
             return ([_("Month"), _("Total")], ["right", "right"])
+    
+    def get_column_header_names(self, engine: Engine) -> (typing.List[str], typing.List[str]):
```
I’d prefer to add a `is_html=True` parameter to `get_column_headers` instead of duplicating code. The addition of a method would also need to be reflected in other reports (VendorStats can’t be rendered as CSV with this code, for example).

Kwpolska commented on expenses/views/reports.py:
```diff
@@ -104,8 +101,18 @@ def report_run(request, slug):
 
     if output_format == "print":
         template = "expenses/report_run_print.html"
+        content_type="text/html"
```
Please use `black` to format your code.

Kwpolska commented on expenses/templates/expenses/report_run_csv.csv:
```diff
@@ -0,0 +1 @@
+    {{ report_html }}
```
Those four spaces appear in the file, which means the first column gets extra four spaces in front. It would be best to remove this file and not use templating altogether when rendering CSVs.

Kwpolska commented on expenses/views/reports.py:
```diff
@@ -118,4 +125,5 @@ def report_run(request, slug):
             "time": end_time - start_time,
             "postfields": postfields,
         },
+        content_type
```
`text/csv` files are downloaded with random filenames by Firefox; a custom filename should be provided to improve user experience.

Kwpolska commented on expenses/views/reports.py:
```diff
@@ -104,8 +101,18 @@ def report_run(request, slug):
+
+    report: Report = report_class(request, settings)
+    start_time = time.monotonic()
+    report_html = report.run(output_format)
```
`report_html` is misleading; it might be better to split the run-and-render part into two functions (for CSV, without using any Django templates; and for HTML).

PrzemyslawSarnacki commented:
Thank you very much for your priceless feedback and sorry for some mess in your code. I did some changes and now csv files are created using csv module. I also deleted unwanted changes and code duplicates. 
Can you elaborate on that quote?  

> this could be fixed by injecting format_money etc. as an argument (and pass a different function depending on output format) 

I'm not really sure how to approach this matter.

Kwpolska commented on expenses/reports.py:
```diff
@@ -118,14 +120,14 @@ def query(self, cursor, sql: str) -> typing.Iterable:
 
     def tabulate(self, results: typing.Iterable, engine: Engine) -> SafeString:
-        column_headers: (typing.List[str], typing.List[str]) = self.get_column_headers(engine)
+        column_headers: (typing.List[str], typing.List[str]) = self.get_column_headers(engine, False)
```
`get_column_headers` should be called with the correct `is_html` value, depending on the output format.

Kwpolska commented on expenses/views/reports.py:
```diff
@@ -86,10 +86,14 @@ def report_run(request, slug):
 
+    report: Report = report_class(request, settings)
+    start_time = time.monotonic()
+    report_html = report.run()
```
Generate only the requested report (HTML or CSV).

Kwpolska commented:
>     this could be fixed by injecting format_money etc. as an argument (and pass a different function depending on output format)

There are two options:

(a) pass `format_money`, `format_number`, `format_link` as arguments to `preprocess_rows`, and pass different functions depending on the output format;  
(b) create a `ReportItemFormatter` class with methods such as `format_money`, `format_number`, `format_link`. Create two subclasses, one for HTML and another for CSV.

I think option (b) is more future-proof.

PrzemyslawSarnacki commented:
> get_column_headers should be called with the correct `is_html` value, depending on the output format.

In this case get_column_headers is called with `is_html = True` value (this is default value).  `tabulate` method is not called during csv creation. 

I made `ReportItemFormatter` classes and implemented proper methods. I've also implemented DailyReports.  I'm looking forward for a feedback.

Kwpolska commented on expenses/reports.py:
```diff
@@ -80,6 +80,39 @@ class TextFieldOption(Option):
+
+class CsvFormatter(ReportItemFormatter):
+
+    """"Remove breaking space from formatted string"""
```
1. A docstring should describe the entire class, not the specific task of a method
2. I would prefer to put just the number directly in the CSV file, maybe formatted according to region / with 2 decimal places (so 1,00 would be fine for Polish, no need for the currency symbol)

Kwpolska commented on expenses/reports.py:
```diff
@@ -80,6 +80,39 @@ class TextFieldOption(Option):
+    def format_category(self, c: Category) -> str:
+        return c
+
+    def format_link(self, vendor: str) -> str:
```
`format_vendor_link` perhaps?

Kwpolska commented on expenses/reports.py:
```diff
@@ -90,7 +123,12 @@ class Report(metaclass=abc.ABCMeta):
     @classmethod
     def meta_to_dict(cls) -> typing.Dict[str, typing.Any]:
-        return {"name": cls.name, "slug": cls.slug, "description": cls.description, "options": cls.options}
+        return {
```
It looks like you ran black with line-length=88; the standard for this project is 120 and `pyproject.toml` should enforce that — please reformat with the correct length.

Kwpolska replied:
(I didn’t really check the rest of the code, because it would require a lot of effort to check if a diff is an actual change or a formatting change.)

PrzemyslawSarnacki replied:
I'm sorry for the issue with black. I accidentally changed my editor's settings and it resulted in wrong formatting

PrzemyslawSarnacki commented:
Formatting seems to be okay now. Sorry for the issue

PrzemyslawSarnacki commented:
I propose that inside `DailySpending` class `run` method should be fragmented into another method (maybe sth like `preprocess_rows` ? that will return `days`, `daily_data`, `user_categories` and 1cat_tables`) so that code duplicates will be avoided. I'm not sure whether this will not result in some side effects.

Kwpolska commented on expenses/reports.py:
```diff
@@ -78,6 +80,40 @@ class TextFieldOption(Option):
+        return format_money(amount)
+
+    def format_category(self, c: Category) -> str:
+        return c
```
This should `return str(c)`, especially since the annotation says so.

Kwpolska commented on expenses/reports.py:
```diff
@@ -78,6 +80,40 @@ class TextFieldOption(Option):
+    """"Subclass for csv items formatting"""
+
+    def format_money(self, amount: typing.Union[int, float, decimal.Decimal]) -> str:
+        return format_number(amount)
```
It might be better to use something that can produce two decimal places for this, and output the correct decimal separator (it could also be configurable).

Kwpolska commented on expenses/reports.py:
```diff
@@ -149,13 +185,42 @@ def tabulate(self, results: typing.Iterable, engine: Engine) -> SafeString:
+        first_row, results = peek(self.preprocess_rows(results, False))
+
+        response = HttpResponse(content_type="text/csv")
+        response["Content-Disposition"] = 'attachment; filename="report.csv"'
```
Or perhaps we could name the file with the report type and current date? (Not necessary, just a bonus suggestion.)

Kwpolska reviewed (changes requested):
Sorry for taking so long to respond, I was a bit busy. There are only some minor things that need changing. It would be great if the field separator (`,`, `;`, `|`, `\t` would be typical options) and number separator (`.` and `,`) were configurable (ust like the currency is), and if money was displayed with the correct number of decimal places — but those are optional, or could be done in a different PR.

Kwpolska commented:
Many thanks for contributing this!
